getnacelle/nacelle-react
# [packages/react-hooks|DD-690] Return promises from asynchronous checkout actions

**Story:** [DD-690](https://nacelle.atlassian.net/browse/DD-690)

> `checkoutActions.processCheckout` promise rejection that can't be handled

### Type of Change


### What is being changed and why?

Asynchronous checkout functions made available in the `checkoutActions` returned from the `useCheckout()` hook were returning `void` instead of returning a Promise. This made it difficult to handle checkout errors when `checkoutActions.processCheckout()` failed for whatever reason.

This PR changes the way that asynchronous `checkoutActions` are handled so that the `getCheckout` and `processChekout` functions return Promises that either:
- reject with a helpful error message
- resolve to an object containing the updated values of `checkoutComplete`, `checkoutId`, `checkoutSource`, and `checkoutUrl`

#### What Else?

- Allows the merchant developer to opt out of automatically redirecting the user to the checkout URL by supplying the prop: `redirectUserToCheckout={false}` to the `<CheckoutProvider />`
  - To protect against breaking changes, the redirect behavior is still true by default 
- Moves asynchronous actions outside of the reducer, which leaves us free to remove the two `dependencies`...`@nacelle/react-hooks` will be dependency-free again!
- Removes unused reducer actions

### How to Test



1. Check out the `DD-690-process-checkout-promise` branch
2. `npm run bootstrap && npm run build`
3. `cd packages/react-hooks && npm run build`
4. `npm run test`
5. Pick an example project, and use the Promise returned by `processCheckout`, both for a regular cart, and for a deliberately malformed cart which should trigger a failure in `processCheckout`. For example, in `examples/nextjs/components/Cart/Cart.js`:

  ```jsx
<Button
  onClick={() =>
    processCheckout({
      lineItems: cart.map((lineItem) => ({
        ...lineItem,
        variant: { ...lineItem.variant, id: '12345' }
      }))
    })
      .then(() => console.log('Success!'))
      .catch((err) => console.warn(`Whoops! ${err}`))
  }
  disabled={!cart.length || isCheckingOut}
  styles={styles.checkoutButton}
  fullwidth={true}
>
  {isCheckingOut ? 'Processing Cart...' : 'Checkout'}
</Button>
  ```

### Demo

#### Handling Checkout Errors
(using the above `<Button />` code in `examples/nextjs/components/Cart/Cart.js`)

https://user-images.githubusercontent.com/5732000/130717009-5558d873-4751-4816-ac3f-6af7276b53a9.mov

The full error message is available in `checkoutState.processCheckoutError`:

<img width="660" alt="demo-process-checkout-error" src="https://user-images.githubusercontent.com/5732000/130717145-02e8a517-ffaa-4b96-a8a5-876aee94b3ae.png">

#### Checkout Success, Without Redirect

https://user-images.githubusercontent.com/5732000/130717077-acfafd10-23c5-4aa6-a8c3-c3a7aba41dea.mov

#### Checkout Success, With Redirect

https://user-images.githubusercontent.com/5732000/130717093-f3a296ed-b153-45a7-a863-d37ae8463a68.mov

### Next Steps

1. Publish `@nacelle/react-hooks@5.1.8` to npm
2. Document changes in [the changelog docs](https://nacelle.atlassian.net/wiki/external/1252819054/OWM0YWYxMzBjM2UxNDdiYmI3ODkzZWVkMjZjYzEzNWY?atlOrigin=eyJpIjoiNWYzNjNiYzY2NDg3NDUxY2I1YmQyZTMzYjgzMTZhM2YiLCJwIjoiYyJ9)
